Use stored procedures to get multi key transactions

arv opened the pull request:
This refactors the code to use stored procedures. This is the way to do
multi key transactions in Cosmos DB.

arv committed:
Use stored procedures to get multi key transactions

This refactors the code to use stored procedures. This is the way to do
multi key transactions in Cosmos DB.

aboodman commented on js/spCreateTodo.js:
```diff
@@ -0,0 +1,55 @@
+  const context = getContext();
+  const collection = context.getCollection();
+
+  // const response = context.getResponse();
```
?

arv replied:
Oh, this file shouild not have been included... let me take another look to see if there is anything else in there

aboodman commented on js/spProcessMutations.js:
```diff
@@ -0,0 +1,119 @@
+async function processMutations(accountID, clientID, mutations) {
+  /** @type {MutationInfo[]} */
+  const infos = [];
+  for (const mutation of mutations) {
```
I think that this creates a longer-lived transaction that is strictly necessary to implement Replicache.

We only need the transaction to span the contents of each iteration of this loop, not the entire loop.

I think this is non-ideal for a sample since people will end up copying it. Can we pull the loop back out to csharp? Sorry.

arv replied:
of course. I misunderstood.

arv committed:
Add updateTodo

aboodman commented on Controllers/BatchController.cs:
```diff
@@ -39,87 +39,57 @@ public async Task<IActionResult> Post(BatchRequest request)
 
             // Process each request in order, ensuring each mutation is
-            // processed in order and exactly once.
+            // processed in order and exactly once.	
```
```suggestion
            // processed in order and exactly once.
```

arv committed:
Add deleteTodo

aboodman commented on Controllers/BatchController.cs:
```diff
@@ -39,87 +39,57 @@ public async Task<IActionResult> Post(BatchRequest request)
-                        Message = e.Message,
-                    });
+                        infos.Add(value as MutationInfo);
+                    }
```
Do you want to panic or something if it's neither of these?

arv replied:
It cannot fail in the current code... but I changed it slightly to make it more resilient

aboodman commented on Db.cs:
```diff
@@ -29,31 +29,59 @@ public Db(IConfiguration configuration)
-                .WithParameter("@id", GetReplicacheStateID(clientID));
-            await foreach (ClientState state in Items.GetItemQueryIterator<ClientState>(def))
+            string sharedPath = "js/shared.js";
+            string utilSource = $"\n// {sharedPath}\n{File.ReadAllText(sharedPath)}";
```
When you run this are you running it on vercel or just locally? I'm kind of surprised that this File API works when deployed on vercel, but I guess it could.

arv replied:
I haven't tried deploying this. `/` paths works everywhere but I guess we might need to use cwd here?

aboodman commented on Model.cs:
```diff
@@ -3,18 +3,33 @@
 namespace todo
 {
-    public class Todo
+    public class Todo<IDType>
```
Can we just store the same type the client wants to send for ID? I may have just used the wrong one.

arv replied:
No. CosmosDB uses the id property for the key in its key-value-store and we want both todo and replicache-client-state to be stored in the same collection.

This is pretty ugly. I was considering how to make this cleaner...

aboodman commented on Db.cs:
```diff
@@ -29,31 +29,59 @@ public Db(IConfiguration configuration)
         }
 
-        public async Task SetMutationID(String accountID, String clientID, UInt64 mutationID)
+        public async Task<UInt64> GetMutationID(String accountID, String clientID)
```
This is not needed anymore, nor is `spGetMutationID`.

arv replied:
This is still needed by the client view

aboodman commented on Db.cs:
```diff
@@ -29,31 +29,59 @@ public Db(IConfiguration configuration)
+            return await ExecuteStoredProcedure<UInt64>(accountID, "spGetMutationID", new dynamic[] { accountID, clientID });
         }
 
         private static String GetReplicacheStateID(String clientID)
```
Is this still needed?

arv replied:
not needed

aboodman commented on js/spGetMutationID.js:
```diff
@@ -0,0 +1,3 @@
+function spGetMutationID(accountID, clientID) {
```
This is not needed anymore.

arv replied:
it is still used

aboodman commented on js/spProcessMutation.js:
```diff
@@ -0,0 +1,135 @@
+// @ts-check
```
Suggestion: we could make this slightly easier to use by refactoring like this:

- `util.js` -> promisification stuff, not replicache specific
- `replicache.js` -> spProcessMutation, getMutationID, setMutationID, getReplicacheStateID but *not* the actual handlers, like `createTodo`
- `mutators.js` -> `createTodo` and friends

This isn't all the way to a real framework, but I think would make the sample easier to copy/paste for users.

arv replied:
Sounds good. I'll do that.

arv committed:
Use unique model classes

arv committed:
Deal with casts and ActionResult

arv committed:
Split the js files

- util.js: Cosmos DB wrappers to make it easier to work with
- mutators.js: All the mutators code
- replicache.js: Other/shared replicache/todo code

We still need the spFoo files because the way CosmosDB stored procedures
work.

arv commented:
I'm going to call this PR done. We can polish this more in new PRs as needed.

aboodman commented on Controllers/BatchController.cs:
```diff
@@ -45,81 +45,51 @@ public async Task<IActionResult> Post(BatchRequest request)
-                    await ProcessMutation(db.Items, accountID, mutation);
+                    return result as BadRequestObjectResult;
                 }
-                catch (PermanentError e)
```
The definition of PermanentError can be removed (and maybe the comment should be moved into JS?)

arv replied:
Good point
